.emacs.d/config.org
#+TITLE: Mark's Emacs Config
#+AUTHOR: Mark Dixon
#+DESCRIPTION: Mark's personal Emacs config.
#+STARTUP: content
#+EXPORT_FILE_NAME: ~/Org/html/config.html
#+OPTIONS: num:nil ^:{}



* ABOUT THIS CONFIG
  Mark's personal Emacs config as a literative / org-moode document. Some useful notes:



* INITIALIZING

Bootstrapping of this config is done in init.el and bootstrap.el - looading
melpa, use-package and some basic stuff.


** Use-package
use-package is a macro that provides convenient syntactic sugar for many
common tasks related to installing and configuring Emacs packages.

#+begin_src emacs-lisp
(straight-use-package 'use-package)
#+end_src



** Keybindings

#+begin_src emacs-lisp

(setq mac-function-modifier 'meta)
(setq mac-command-modifier 'meta)
(setq mac-option-modifier nil)

(global-set-key (kbd "C-z") 'undo)  ; nice to be a bit lazy

#+end_src


** Sane defaults
From old Emacs config - these are the small settings that make a nice difference
to the whole Emacs experience..

#+begin_src emacs-lisp

(setq-default frame-title-format '("%b")) ; Make window title the buffer name
(fset 'yes-or-no-p 'y-or-n-p)             ; y-or-n-p makes answering questions faster
(delete-selection-mode 1)                 ; Selected text will be overwritten when you start typing
(setq use-dialog-box nil)                 ;; Prompts should go in the minibuffer, not in a GUI.
(setq-default indent-tabs-mode nil)       ; never use tabs!
(set-charset-priority 'unicode)           ; prioritise utf-8
(setq locale-coding-system 'utf-8)
(set-terminal-coding-system 'utf-8)
(set-keyboard-coding-system 'utf-8)
(set-selection-coding-system 'utf-8)
(prefer-coding-system 'utf-8)
(setq default-process-coding-system '(utf-8-unix . utf-8-unix))
(global-auto-revert-mode t)               ; Auto-update buffer if file has changed on disk

#+end_src

#+RESULTS:
: t


This one is important - it adds line wrapping which makes org-mode generally nicer.

#+begin_src emacs-lisp
(global-visual-line-mode t)
#+end_src


** Icons
This is an icon set that can be used with dashboard, dired, ibuffer and other Emacs programs.

All-the-icons seems to be the only real icon set being updated for Emacs.

#+begin_src emacs-lisp
(use-package all-the-icons
  :straight t)
#+end_src


** Emojis
Emojify is an Emacs extension to display emojis. It can display github style emojis like :smile: or plain ascii ones like :).

#+Begin_src emacs-lisp
(use-package emojify
  :straight t
  :hook (after-init . global-emojify-mode))
#+end_src





** Spellchecking


#+begin_src emacs-lisp

(setq ispell-program-name "/opt/homebrew/bin/ispell")

(use-package guess-language         ; Automatically detect language for Flyspell
  :ensure t
  :defer t
  :straight t
  :init (add-hook 'org-mode-hook #'guess-language-mode)
  :config
  (setq guess-language-langcodes '((en . ("en_GB" "English"))
                                   (sv . ("sv" "Swedish")))
        guess-language-languages '(en sv)
        guess-language-min-paragraph-length 45)
  :diminish guess-languag
  e-mode)
#+end_src

*** Jinx

#+begin_src emacs-lisp

(use-package jinx
  :ensure t
  :straight t
  :config
  (setq jinx-languages '("en_GB"))
  (global-jinx-mode)
  )

#+end_src

#+RESULTS:
: t



* NANO EMACS

Load Nano-Emacs and Nano-Theme via straight... directly from Github.

#+begin_src emacs-lisp
;; (straight-use-package
;;   '(nano-emacs :type git :host github :repo "rougier/nano-emacs"))

;; (straight-use-package
;;  '(nano-emacs :type git :host github :repo "rougier/nano-emacs"))

(straight-use-package
  '(nano-emacs :type git :host github :repo "rougier/nano-emacs"
               :fork (:host github
                    :repo "wakatara/nano-emacs")))

#+end_src

  
** Nano Layout
This module defines the overall layout of an emacs frame, defining default font, fringes, margins, etc.

#+begin_src emacs-lisp
;; Default layout (optional)
(require 'nano-layout)
#+end_src

** Nano Fonts

#DEPR+begin_src emacs-lisp

(set-face-attribute 'default nil
  :font "Hack Nerd Font"
  :height 150
  :weight 'light)
(set-face-attribute 'variable-pitch nil
  :font "SF Pro Text"
  :height 180
  :weight 'medium)
(set-face-attribute 'fixed-pitch nil
  :font "Hack Nerd Font"
  :height 150
  :weight 'medium)

;; Makes commented text and keywords italics.
;; This is working in emacsclient but not emacs.
;; Your font must have an italic face available.
(set-face-attribute 'font-lock-comment-face nil
  :slant 'italic)
(set-face-attribute 'font-lock-keyword-face nil
  :slant 'italic)

;; Uncomment the following line if line spacing needs adjusting.
(setq-default line-spacing 0)

;; Needed if using emacsclient. Otherwise, your fonts will be smaller than expected.
;;(add-to-list 'default-frame-alist '(font . "Hack-14"))


#+end_src

#+RESULTS:
: SF Pro Text

#+begin_src emacs-lisp
(setq nano-font-family-monospaced "Hack Nerd Font")
(setq nano-font-family-proportional "SF Pro Text")
(setq nano-font-size 16)
#+end_src

#+RESULTS:
: 16

** Nano Theme
This module derives faces for several popular emacs modes from the nano faces.

Currently loading Nord theme before nano theme in order to give nicer colorations
to org-roam.

#+begin_src emacs-lisp
(straight-use-package
    '(nano-theme :type git :host github :repo "rougier/nano-theme"))


;;(require 'nano-theme-dark)
;;(nano-theme-set-dark)
;;(require 'nano-theme)
;;(nano-theme)
;;(nano-light)
;;(nano-dark)

(set-face-attribute 'default nil :family "Fira Code" :weight 'normal :height 160)
;;(set-face-attribute 'default nil :weight 'normal :height 160)

(use-package nano-theme
    :ensure t
    :straight t
    :config
    (custom-set-faces
        '(italic ((t (:inherit nano-salient))))     
        '(org-level-2 ((t (:inherit nano-default))))
        '(org-level-3 ((t (:inherit nano-default))))
        '(org-level-4 ((t (:inherit nano-default))))
        '(org-level-5 ((t (:inherit nano-default))))
        '(org-level-6 ((t (:inherit nano-default))))
        '(org-level-7 ((t (:inherit nano-default))))
        '(org-level-8 ((t (:inherit nano-default))))
        '(org-table ((t (:inherit nano-default)))))
  )
(load-theme 'nano-dark t)



;;(defface nano-mono
;;  '((t (:family "Roboto Mono"
;;        :height 140
;;        :weight light)))
;;  "Default monospaced font (Roboto Mono Light, 14pt)."
;;  :group 'nano-theme-fonts)

;;    (set-face 'org-headline-done                       'nano-face-faded)
;;    (set-face 'org-level-1                            'nano-face-strong)
;;    (set-face 'org-level-2                            'nano-face-strong)
;;    (set-face 'org-level-3                            'nano-face-strong)
;;    (set-face 'org-level-4                            'nano-face-strong)
;;    (set-face 'org-level-5                            'nano-face-strong)
;;    (set-face 'org-level-6                            'nano-face-strong)
;;    (set-face 'org-level-7                            'nano-face-strong)
;;    (set-face 'org-level-8                            'nano-face-strong)
;;    (set-face 'org-upcoming-deadline                 'nano-face-default)


;; Theme
(require 'nano-faces)
(nano-faces)



#+end_src

#+RESULTS:


** Nano Defaults
This modules provides a set of sane default settings for Emacs.

#+begin_src emacs-lisp
;; Nano default settings (optional)
(require 'nano-defaults)

;; Default shell in term
(unless (eq system-type 'windows-nt)
  (setq-default shell-file-name "/usr/local/bin/fish")
  (setq explicit-shell-file-name "/usr/local/bin/fish"))

;; nano-defaults sets windmove default to shift - but this is not nice with org-mode so set it to hyper instead
(windmove-default-keybindings 'hyper)
  #+end_src


** Nano Colors
This module provides a collection of colors palettes (open colors, material colors, nord colors) with functions for easily accessing them.

#+begin_src emacs-lisp
;; Nano colors (optional)
(require 'nano-colors)
#+end_src


** Nano Session
This modules configures Emacs such that a session is saved from one run to the other.

#+begin_src emacs-lisp
;; Nano session saving (optional)
(require 'nano-session)
#+end_src


** Nano Modeline
This module defines a header line that is mode dependent and takes care
of hiding the modeline when necessary.

#+begin_src emacs-lisp
;; Nano header & mode lines (optional)
(require 'nano-modeline)
#+end_src

#+RESULTS:
: nano-modeline


** Nano Splash
This module provides a splash screen when emacs is started.

#+begin_src emacs-lisp
;; Splash (optional)
(unless (member "-no-splash" command-line-args)
    (require 'nano-splash))
#+end_src


** Nano Help
This module provides a function to display a small help message in the
echo area.

#+begin_src emacs-lisp
;; Help (optional)
(unless (member "-no-help" command-line-args)
    (require 'nano-help))
#+end_src

#+RESULTS:
: nano-help

** Nano MU4E
#+begin_src emacs-lisp
;;(use-package svg-tag-mode
;;  :straight t)

;;(straight-use-package
;; '(mu4e-thread-folding :type git :host github :repo "rougier/mu4e-thread-folding"))
;;(straight-use-package
;; '(mu4e-dashboard :type git :host github :repo "rougier/mu4e-dashboard"))


;;(require 'nano-mu4e)
#+end_src

#+RESULTS:
: t



* PROGRAMMING + Tree-Sitter
** Install module

#+begin_src emacs-lisp
(require 'treesit)
#+end_src

#+RESULTS:
: treesit

** Python settings

#+begin_src elisp
(defvar python--treesit-settings
  (treesit-font-lock-rules
   :feature 'comment
   :language 'python
   '((comment) @font-lock-comment-face)

   :feature 'string
   :language 'python
   '((string) @font-lock-string-face
     (string) @contextual) ; Contextual special treatment.

   :feature 'function-name
   :language 'python
   '((function_definition
      name: (identifier) @font-lock-function-name-face))

   :feature 'class-name
   :language 'python
   '((class_definition
      name: (identifier) @font-lock-type-face))

   ))
#+end_src

#+RESULTS:
: python--treesit-settings

#+begin_src emacs-lisp
(set-face-attribute 'font-lock-function-name-face nil :slant 'italic)

#+end_src

#+RESULTS:

#+begin_src elisp
(define-derived-mode python-mode prog-mode "Python"
  (cond
   ;; Tree-sitter.
   ((treesit-ready-p 'python-mode 'python)
    (treesit-parser-create 'python)
    (setq-local treesit-font-lock-settings python--treesit-settings)
    (setq-local treesit-font-lock-feature-list
                '((comment string function-name)
                  (class-name keyword builtin)
                  (string-interpolation decorator)))
    (treesit-major-mode-setup))
   (t
    ;; No tree-sitter
    ;;(setq-local font-lock-defaults ...)
    )))
#+end_src

#+RESULTS:
: python-mode





** VTERM

#+begin_src emacs-lisp
(use-package vterm
  :straight t
  :ensure t)

#+end_src

#+RESULTS:

** Julia REPL

#+begin_src emacs-lisp

(use-package julia-mode
  :straight t
  :ensure t)

(use-package julia-repl
  :straight t
  :ensure t
  :hook (julia-mode . julia-repl-mode)

  :init
  (setenv "JULIA_NUM_THREADS" "8")

  :config
  ;; Set the terminal backend
  (julia-repl-set-terminal-backend 'vterm)
  
  ;; Keybindings for quickly sending code to the REPL
  (define-key julia-repl-mode-map (kbd "<C-RET>") 'my/julia-repl-send-cell)
  (define-key julia-repl-mode-map (kbd "<M-RET>") 'julia-repl-send-line)
  (define-key julia-repl-mode-map (kbd "<S-return>") 'julia-repl-send-buffer))

#+end_src

#+RESULTS:
| julia-repl-mode |

And a function to send code "cells" to the REPL..

#+begin_src emacs-lisp
(defun my/julia-repl-send-cell() 
  ;; "Send the current julia cell (delimited by ###) to the julia shell"
  (interactive)
  (save-excursion (setq cell-begin (if (re-search-backward "^###" nil t) (point) (point-min))))
  (save-excursion (setq cell-end (if (re-search-forward "^###" nil t) (point) (point-max))))
  (set-mark cell-begin)
  (goto-char cell-end)
  (julia-repl-send-region-or-line)
  (next-line))
#+end_src

#+RESULTS:
: my/julia-repl-send-cell


* UTIL PACKAGES

** Smex
Smex is a package the makes M-x remember our history.  Now M-x will show our
last used commands first.

#+begin_src emacs-lisp
;;(straight-use-package 'smex)
(use-package smex
  :straight t)
(smex-initialize)
#+end_src


** WRITEROOM MODE
A minor mode for Emacs that implements a distraction-free writing mode similar to the famous Writeroom editor for OS X.

#+begin_src emacs-lisp
(use-package writeroom-mode
    :config (setq writeroom-width 80)
    :straight t)
#+end_src


  
 

* ORG MODE

#+begin_src emacs-lisp
(use-package org
  :ensure org-contrib
  ;; The rest of your org-mode configuration
)
#+end_src

#+RESULTS:

** Defining A Few Things

#+begin_src emacs-lisp
(add-hook 'org-mode-hook 'org-indent-mode)
(add-hook 'org-mode-hook 'turn-on-flyspell)

(setq org-directory "~/Dropbox/org/"
      org-agenda-files '("~/Dropbox/org-roam/")
      org-default-notes-file (expand-file-name "notes.org" org-directory)
      org-ellipsis " [+] "
      org-startup-indented t
      org-log-done 'time
      org-return-follows-link  t
      org-journal-dir "~/Dropbox/org-roam/daily/"
      org-journal-date-format "%Y-%m-%d (%A) "
      org-journal-file-format "%Y-%m-%d.org"
      org-hide-emphasis-markers nil)
;; (setq org-ellipsis "⚡⚡⚡");; ⤵
(setq org-src-preserve-indentation nil
      org-src-tab-acts-natively t
      org-edit-src-content-indentation 0)

;; links open in same frame
(setq org-link-frame-setup '((file . find-file)))


;; better refile
(setq org-refile-targets '((nil :maxlevel . 9)
                                (org-agenda-files :maxlevel . 9)))
(setq org-outline-path-complete-in-steps nil)         ; Refile in a single go
(setq org-refile-use-outline-path t)                  ; Show full paths for refiling



#+end_src


** Org agenda customisations

Full window

#+begin_src emacs-lisp
(setq org-agenda-window-setup 'current-window)
#+end_src


** Org Speed Commands

#+begin_src emacs-lisp
;; org speed key commands
(setq org-use-speed-commands t)
(setq org-speed-commands (cons '("w" . widen) org-speed-commands))
#+end_src

#+RESULTS:
: ((w . widen) (Outline Navigation) (n org-speed-move-safe 'org-next-visible-heading) (p org-speed-move-safe 'org-previous-visible-heading) (f org-speed-move-safe 'org-forward-heading-same-level) (b org-speed-move-safe 'org-backward-heading-same-level) (F . org-next-block) (B . org-previous-block) (u org-speed-move-safe 'outline-up-heading) (j . org-goto) (g org-refile '(4)) (Outline Visibility) (c . org-cycle) (C . org-shifttab) (  . org-display-outline-path) (s . org-toggle-narrow-to-subtree) (k . org-cut-subtree) (= . org-columns) (Outline Structure Editing) (U . org-metaup) (D . org-metadown) (r . org-metaright) (l . org-metaleft) (R . org-shiftmetaright) (L . org-shiftmetaleft) (i progn (forward-char 1) (call-interactively 'org-insert-heading-respect-content)) (^ . org-sort) (w . org-refile) (a . org-archive-subtree-default-with-confirmation) (@ . org-mark-subtree) (# . org-toggle-comment) (Clock Commands) (I . org-clock-in) (O . org-clock-out) (Meta Data Editing) (t . org-todo) (, org-priority) (0 org-priority 32) (1 org-priority 65) (2 org-priority 66) (3 org-priority 67) (: . org-set-tags-command) (e . org-set-effort) (E . org-inc-effort) (W lambda (m) (interactive sMinutes before warning: ) (org-entry-put (point) APPT_WARNTIME m)) (Agenda Views etc) (v . org-agenda) (/ . org-sparse-tree) (Misc) (o . org-open-at-point) (? . org-speed-command-help) (< org-agenda-set-restriction-lock 'subtree) (> org-agenda-remove-restriction-lock))

** Nicer org bullets with org-bars

#+begin_src emacs-lisp

(straight-use-package
 '(org-bars :type git :host github :repo "tonyaldon/org-bars"))

(require 'org-bars)
(add-hook 'org-mode-hook #'org-bars-mode)

(setq org-bars-stars '(:empty "*"
                       :invisible "*"
                       :visible "*"))

(setq org-bars-color-options '(:only-one-color t
                               :bar-color "#434c5e"))


#+end_src

#+RESULTS:
| :only-one-color | t | :bar-color | #434c5e |


** Org Todo Keywords
This lets us create the various TODO tags that we can use in Org.

#+begin_src emacs-lisp
  (setq org-todo-keywords        ; This overwrites the default Doom org-todo-keywords
          '((sequence
             "TODO(t)"           ; A task that is ready to be tackled
             "PROJ(p)"           ; A project that contains other tasks
             "DOING(g)"           ; Something is holding up this task
             "STALLED(w)"           ; Something is holding up this task
             "TESTING(t)"           ; Something is holding up this task
             "|"                 ; The pipe necessary to separate "active" states and "inactive" states
             "DONE(d)"           ; Task has been completed
             "CANCELLED(c)" )))  ; Task has been cancelled


(setq org-todo-keyword-faces
  '(
    ("TODO" . (:foreground "#b48ead" :weight bold))
    ("DOING" . "#a3be8c")
    ("PROJ" . "#d08770")
    ("DONE" . "#8fbcbb")
    ("CANCELLED" . "#bf616a")
   )
)

(setq org-fontify-done-headline t)

(custom-set-faces
 '(org-headline-done
            ((((class color) (min-colors 16) (background dark))
               (:foreground "dim gray" :bold nil
               :strike-through nil))))
)


#+end_src

#+RESULTS:




** Source Code Block Tag Expansion
Org-tempo is a package that allows for '<s' followed by TAB to expand to a begin_src tag.  Other expansions available include:


| Typing the below + TAB | Expands to ...                          |
|------------------------+-----------------------------------------|
| <a                     | '#+BEGIN_EXPORT ascii' … '#+END_EXPORT  |
| <c                     | '#+BEGIN_CENTER' … '#+END_CENTER'       |
| <C                     | '#+BEGIN_COMMENT' … '#+END_COMMENT'     |
| <e                     | '#+BEGIN_EXAMPLE' … '#+END_EXAMPLE'     |
| <E                     | '#+BEGIN_EXPORT' … '#+END_EXPORT'       |
| <h                     | '#+BEGIN_EXPORT html' … '#+END_EXPORT'  |
| <l                     | '#+BEGIN_EXPORT latex' … '#+END_EXPORT' |
| <q                     | '#+BEGIN_QUOTE' … '#+END_QUOTE'         |
| <s                     | '#+BEGIN_SRC' … '#+END_SRC'             |
| <v                     | '#+BEGIN_VERSE' … '#+END_VERSE'         |

#+begin_src emacs-lisp
(use-package org-tempo
  :ensure nil) ;; tell use-package not to try to install org-tempo since it's already there.
#+end_src


*** Easier code bock for literate config

While writing this configuration file in Org mode, I have to write code
blocks all the time. Org has templates, so doing =<s TAB= creates a source
code block. Here I create a custom template for emacs-lisp specifically.
So, =<el TAB= creates the Emacs lisp code block and puts the cursor inside.


  #+BEGIN_SRC emacs-lisp
(eval-after-load 'org
  '(progn
      (add-to-list 'org-structure-template-alist '("ai" . "ai :maxtokens 512 :temperature 0.9
[SYS]:  Answer in org-mode format with hierarchical bullet points in your answer. 

[ME]:"))
      (add-to-list 'org-structure-template-alist '("img" . "ai :image "))
      (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
      (add-to-list 'org-structure-template-alist '("py" . "src python"))
      (add-to-list 'org-structure-template-alist '("bash" . "src bash"))
   )
)
  #+END_SRC

  #+RESULTS:
  : ((aid . ai :maxtokens 2000 :temperature 0.9 
  : [SYS]: You are an AI assistant. 
  : [ME]:) (aim . ai :maxtokens 2000 :temperature 0.9 
  : [SYS]: You are an AI assistant. 
  : [ME]: ) (py . src python) (el . src emacs-lisp) (img . ai :image ) (ai . ai :maxtokens 2000 :temperature 0.9 ) (a . export ascii) (c . center) (C . comment) (e . example) (E . export) (h . export html) (l . export latex) (q . quote) (s . src) (v . verse))

  
  
** Source Code Block Syntax Highlighting
We want the same syntax highlighting in source blocks as in the native language files.

#+begin_src emacs-lisp
(setq org-src-fontify-natively t
    org-src-tab-acts-natively t
    org-confirm-babel-evaluate nil
    org-edit-src-content-indentation 0)

(org-babel-do-load-languages
 'org-babel-load-languages

.config/qtile/config.py
# Copyright (c) 2010 Aldo Cortesi
# Copyright (c) 2010, 2014 dequis
# Copyright (c) 2012 Randall Ma
# Copyright (c) 2012-2014 Tycho Andersen
# Copyright (c) 2012 Craig Barnes
# Copyright (c) 2013 horsik
# Copyright (c) 2013 Tao Sauvage
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from typing import List  # noqa: F401

from libqtile import bar, layout, widget
from libqtile.config import Click, Drag, Group, Key, Match, Screen, ScratchPad, DropDown
from libqtile.lazy import lazy

from libqtile.utils import guess_terminal
from libqtile.log_utils import logger
from libqtile import qtile
import libqtile

mod = "mod4"
terminal = guess_terminal()



keys = [
    # Switch between windows
    Key([mod], "Left", lazy.layout.left(), desc="Move focus to left"),
    Key([mod], "Right", lazy.layout.right(), desc="Move focus to right"),
    Key([mod], "Down", lazy.layout.down(), desc="Move focus down"),
    Key([mod], "Up", lazy.layout.up(), desc="Move focus up"),
    Key([mod], "space", lazy.layout.next(),
        desc="Move window focus to other window"),

    # Move windows between left/right columns or move up/down in current stack.
    # Moving out of range in Columns layout will create new column.
    Key([mod, "shift"], "Left", lazy.layout.shuffle_left(),
        desc="Move window to the left"),
    Key([mod, "shift"], "Right", lazy.layout.shuffle_right(),
        desc="Move window to the right"),
    Key([mod, "shift"], "Down", lazy.layout.shuffle_down(),
        desc="Move window down"),
    Key([mod, "shift"], "Up", lazy.layout.shuffle_up(), desc="Move window up"),

    # Grow windows. If current window is on the edge of screen and direction
    # will be to screen edge - window would shrink.
    Key([mod, "control"], "Left", lazy.layout.grow_left(),
        desc="Grow window to the left"),
    Key([mod, "control"], "Right", lazy.layout.grow_right(),
        desc="Grow window to the right"),
    Key([mod, "control"], "Down", lazy.layout.grow_down(),
        desc="Grow window down"),
    Key([mod, "control"], "Up", lazy.layout.grow_up(), desc="Grow window up"),
    Key([mod], "n", lazy.layout.normalize(), desc="Reset all window sizes"),



    # Toggle between split and unsplit sides of stack.
    # Split = all windows displayed
    # Unsplit = 1 window displayed, like Max layout, but still with
    # multiple stack panes
    Key([mod, "shift"], "Return", lazy.layout.toggle_split(),
        desc="Toggle between split and unsplit sides of stack"),
    Key([mod], "Return", lazy.spawn(terminal), desc="Launch terminal"),

    # Toggle between different layouts as defined below
    Key([mod], "Tab", lazy.next_layout(), desc="Toggle between layouts"),
    Key([mod, "shift"], "q", lazy.window.kill(), desc="Kill focused window"),

    Key([mod], "r", lazy.reload_config(), desc="Reload Qtile config"),
    Key([mod, "shift"], "r", lazy.restart(), desc="Restart Qtile"),
    Key([mod, "control"], "q", lazy.shutdown(), desc="Shutdown Qtile"),
    #Key([mod], "r", lazy.spawncmd(), desc="Spawn a command using a prompt widget"),
    # Rofi and dmenu:
    Key([mod], "d", lazy.spawn("rofi -show drun -config /home/mark/.config/rofi/Nord.rasi"), desc="Rofi"),
    Key([mod, "shift"], "d", lazy.spawn("dmenu_run -i -nb '#2E3440' -nf '#81a1c1' -sb '#81a1c1' -sf '#2E3440' -fn 'SFDisplayPro:bold:pixelsize=14'"), desc="Rofi"),

    Key([mod], "F12",
        lazy.group["scratchpad"].dropdown_toggle('term'),
        desc='Dropdown terminal'
    ),

    Key([mod], "f",
        lazy.window.toggle_fullscreen(),
        desc='toggle fullscreen'
    ),

    Key([mod, "shift"], "space",
        lazy.window.toggle_floating(),
        desc='toggle floating'
    ),
    

]

groups = [
    #Group(i) for i in "123456789" +
    #  
    
    Group('1', label='', layout='columns'),
    Group('2', label='', layout='columns'),
    Group('3', label='', layout='columns'),
    Group('4', label='', layout='columns'),
    Group('5', label='', layout='columns'),
    Group('6', label='', layout='columns'),
    Group('7', label='', layout='columns'),
    Group('8', label='', layout='columns'),
    Group('9', label='﫯', layout='columns'),
    Group('0', label='', layout='columns'),
    
    ScratchPad("scratchpad", [
        # define a drop down terminal.
        # it is placed in the upper third of screen by default.
        DropDown("term", "alacritty -e R -q --no-save", opacity=0.95),

    ] ),
]

for i in groups[:-1]:
    keys.extend([
        # mod1 + letter of group = switch to group
        Key([mod], i.name, lazy.group[i.name].toscreen(),
            desc="Switch to group {}".format(i.name)),

        # mod1 + shift + letter of group = switch to & move focused window to group
        Key([mod, "shift"], i.name, lazy.window.togroup(i.name, switch_group=True),
            desc="Switch to & move focused window to group {}".format(i.name)),
        # Or, use below if you prefer not to switch to that group.
        # # mod1 + shift + letter of group = move focused window to group
        # Key([mod, "shift"], i.name, lazy.window.togroup(i.name),
        #     desc="move focused window to group {}".format(i.name)),
    ])





########################
# Define colors ########
########################

colors = [
    ["#2e3440", "#2e3440"],  # background 0
    ["#d8dee9", "#d8dee9"],  # foreground 1
    ["#4c566a", "#4c566a"],  # background lighter 2
    ["#bf616a", "#bf616a"],  # red 3
    ["#a3be8c", "#a3be8c"],  # green 4
    ["#ebcb8b", "#ebcb8b"],  # yellow 5
    ["#81a1c1", "#81a1c1"],  # blue 6
    ["#b48ead", "#b48ead"],  # magenta 7
    ["#88c0d0", "#88c0d0"],  # cyan 8
    ["#eceff4", "#eceff4"],  # white 9
    ["#4c566a", "#4c566a"],  # grey 10
    ["#d08770", "#d08770"],  # orange 11
    ["#8fbcbb", "#8fbcbb"],  # super cyan 12
    ["#88c0d0", "#88c0d0"],  # super blue 13
    ["#111318", "#111318"],  # super dark background 14
]

layout_theme = {
    "border_width": 2,
    "margin": 3,
    "border_on_single": False,
    "margin_on_single": [ 3, 0, 0, 0 ],
    "fair": True, # Add new windows to the column with least windows.
    "border_focus": "5e81ac",
    "border_normal": "4c566a",
    "font": "Hack Nerd Font",
    "grow_amount": 2,
}


layouts = [
    #layout.Columns(border_focus_stack=['#d75f5f', '#8f3d3d'], border_width=4),
    layout.Columns(**layout_theme),
    layout.Max(**layout_theme),
    # layout.Floating(**layout_theme),

    # Try more layouts by unleashing below layouts.
    # layout.Stack(num_stacks=2),
    # layout.Bsp(),
    # layout.Matrix(),
    # layout.MonadTall(),
    # layout.MonadWide(),
    # layout.RatioTile(),
    # layout.Tile(),
    # layout.TreeTab(),
    # layout.VerticalTile(),
    # layout.Zoomy(),
]

widget_defaults = dict(
    font='SF Pro Display',
    #font='Hack Nerd Font',
    fontsize=14,
    padding=3,
    background=colors[0],
)
extension_defaults = widget_defaults.copy()


def separator(linewidth=1, padding=6):
    return widget.Sep(
        linewidth = linewidth,
        padding = padding,
        foreground = colors[2],
        background = colors[0]
    )



screens = [
    Screen(
        top=bar.Bar( [
            widget.CurrentLayoutIcon( scale=0.6 ),
            widget.GroupBox(
                font = "SF Pro Display",
                fontsize = 24,
                margin_y = 3,
                margin_x = 0,
                padding_y = 5,
                padding_x = 3,
                borderwidth = 3,
                active = colors[9],
                inactive = colors[2],
                this_current_screen_border = colors[6],
                rounded = False,
                disable_drag=True,
                highlight_color = colors[1],
                highlight_method = "block", # "block"
            ),
            widget.Spacer(),
            widget.WindowName(
                width=bar.CALCULATED,
                font = "SF Pro Display",
                #font = "Hack Nerd Pro",
                fontsize = 14,
                foreground=colors[6],
            ),
            widget.Spacer(),
            separator(0),
            widget.CPUGraph(
                mouse_callbacks={
                    'Button1': lambda: qtile.cmd_spawn('alacritty -e bpytop')
                }
            ),
            separator(),
            widget.BatteryIcon(),
            widget.Battery(),
            separator(),
            widget.PulseVolume(),
            separator(),
            widget.Wttr(
                lang='en',
                location={
                    'Stockholm': 'Stockholm',
                    },
                format='%m %t',
                units='m',
                update_interval=30,
                mouse_callbacks={
                    'Button1': lambda: qtile.current_screen.set_group(qtile.groups[2]) or qtile.cmd_spawn('firefox https://moonphases.co.uk/') 
                }
            ),
            separator(),
            widget.TextBox('', fontsize=24),
            widget.Clock(
                format='%-H:%M',
                mouse_callbacks={
                    'Button1': lambda: qtile.current_screen.toggle_group(qtile.groups[9],warp=False)
                },
            ),
            #separator(),
            #widget.Clock(
            #    format='%Y-%m-%d',
            #    mouse_callbacks={
            #        'Button1': lambda: qtile.current_screen.toggle_group(qtile.groups[9],warp=False)
            #    },
            #),
            separator(),
            widget.WidgetBox(
                widgets=[
                    widget.Systray(),                     
                    ],
                font="Hack Nerd Font",
                fontsize=14,
                text_open="   ",
                text_closed="   ",
                close_button_location="right"
            ),
            separator(),
            widget.QuickExit(
                default_text="襤",
                countdown_format='{}',
                font="Hack Nerd Font",
                fontsize=14,
            ),
            separator(0)
            ],
            22,
            opacity=0.9
        ),
    ),
    ]

# Drag floating layouts.
mouse = [
    Drag([mod], "Button1", lazy.window.set_position_floating(),
         start=lazy.window.get_position()),
    Drag([mod], "Button3", lazy.window.set_size_floating(),
         start=lazy.window.get_size()),
    Click([mod], "Button2", lazy.window.bring_to_front())
]

dgroups_key_binder = None
dgroups_app_rules = []  # type: List
follow_mouse_focus = True
bring_front_click = False
cursor_warp = False
floating_layout = layout.Floating(float_rules=[
    # Run the utility of `xprop` to see the wm class and name of an X client.
    *layout.Floating.default_float_rules,
    Match(wm_class='confirmreset'),  # gitk
    Match(wm_class='makebranch'),  # gitk
    Match(wm_class='maketag'),  # gitk
    Match(wm_class='ssh-askpass'),  # ssh-askpass
    Match(title='branchdialog'),  # gitk
    Match(title='pinentry'),  # GPG key password entry
])
auto_fullscreen = True
focus_on_window_activation = "smart"
reconfigure_screens = True

# If things like steam games want to auto-minimize themselves when losing
# focus, should we respect this or not?
auto_minimize = True

# XXX: Gasp! We're lying here. In fact, nobody really uses or cares about this
# string besides java UI toolkits; you can see several discussions on the
# mailing lists, GitHub issues, and other WM documentation that suggest setting
# this string if your java app doesn't work correctly. We may as well just lie
# and say that we're a working one by default.
#
# We choose LG3D to maximize irony: it is a 3D non-reparenting WM written in
# java that happens to be on java's whitelist.
wmname = "LG3D"


#lazy.restart()



@libqtile.hook.subscribe.client_new
def func(c):
    #logger.warning( "%s = %s" % ( c.name, c.window.get_wm_class() ) )
    if c.name == "Mailspring":
        c.togroup("9")
    elif "Emacs" in c.window.get_wm_class():
        c.togroup("2")
    elif "firefox" in c.window.get_wm_class():
        c.togroup("3")
    elif "Morgen" in c.window.get_wm_class():
        c.togroup("0")
    elif "Code" in c.window.get_wm_class():
        c.togroup("5")
    elif "Spotify" in c.window.get_wm_class():
        c.togroup("7")
    elif "zoom" in c.window.get_wm_class():
        c.togroup("8")
    elif c.name == "dzen":
        c.cmd_static(0)

